dask/array/core.py

```diff
@@ -2009,6 +2019,83 @@ def from_array(x, chunks, name=None, lock=False, asarray=True, fancy=True,
 
 
+def from_zarr(url, component=None, storage_options=None, **kwargs):
+    """Load array from the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    url: str
+        Location of the data. This can include a protocol specifier like s3://
+        for remote data.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be loaded, something like ``'foo/bar'``.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    kwargs: passed to zarr's open functions.
+    """
+    import zarr
+    storage_options = storage_options or {}
+    fs, fs_token, path = get_fs_token_paths(
+        url, 'rb', storage_options=storage_options)
+    assert len(path) == 1
+    mapper = get_mapper(fs, path[0])
+    if component is None:
+        z = zarr.open_array(mapper, mode='r', **kwargs)
+    else:
+        z = zarr.open_group(mapper, mode='r', **kwargs)[component]
+    return from_array(z, z.chunks, name='zarr-%s' % url)
+
+
+def to_zarr(arr, url, component=None, storage_options=None,
+            overwrite_group=False, compute=True, return_stored=False, **kwargs):
+    """Save array to the zarr storage format
+
+    See https://zarr.readthedocs.io for details about the format.
+
+    Parameters
+    ----------
+    arr: dask.array
+        Data to store
+    url: str
+        Location of the data. This can include a protocol specifier like s3://
+        for remote data.
+    component: str or None
+        If the location is a zarr group rather than an array, this is the
+        subcomponent that should be created/over-written.
+    storage_options: dict
+        Any additional parameters for the storage backend (ignored for local
+        paths)
+    overwrite_group: bool
+        If component is not none, this controls whether to create/overwrite the
+        zarr group, or if the group must exits already (but the specific
+        dataset is always overwritten)
+    compute, return_stored: see ``store()``
+    kwargs: passed to zarr's open functions, e.g., compression options
+    """
+    import zarr
+    # TODO: rechunk here, if necessary
+    storage_options = storage_options or {}
+    fs, fs_token, path = get_fs_token_paths(
+        url, 'rb', storage_options=storage_options)
+    assert len(path) == 1
+    mapper = get_mapper(fs, path[0])
+    chunks = [c[0] for c in arr.chunks]
+    if component is None:
+        z = zarr.open_array(
+            mapper, mode='w', shape=arr.shape, chunks=chunks, dtype=arr.dtype,
+            **kwargs)
+    else:
+        mode = 'w' if overwrite_group else 'r+'
+        z = zarr.open_group(mapper, mode=mode).create_dataset(
+            component, shape=arr.shape, chunks=chunks, dtype=arr.dtype,
+            **kwargs)
+    return store(arr, z, compute=compute, return_stored=return_stored)
```

Comment:
Though we might want to provide the option to specify a `lock` for writing.

Reply:
If the dask array to be stored has regular chunks, then I think a lock is not needed, because the writes will be aligned with the chunks in the newly created zarr array. In that case could probably pass ``lock=False`` to ``store()``. 

Not sure what to do if dask array has irregular chunks. Storing should still be possible but writes may not be aligned with zarr array chunks. Is it worth doing something like ``arr = rechunk(arr, z.chunks)`` just to be sure writes will be aligned?

Reply:
It sounds like we should `rechunk` before calling store?  Presumably it's better to rechunk beforehand.

Reply:
Actually, it might be nice to centralize that logic within `store` itself with an optional `chunks=` parameter.  Presumably we might choose to align dask.array chunks so that they align with, but are possibly larger than, the given chunks

Reply:
I think that's exactly wha I suggested in the chunks issue. Here we should apply rechunk as given, including the option of not rechunking, and error if the chunks are not regular.

Reply:
Agreed using `rechunk` is much better than locking (that is what we do).

As to how to irregular chunks should be dealt with, there is some discussion in issue ( https://github.com/dask/dask/issues/3302 ) about how to do this. Would be curious to hear your thoughts over there, @alimanfoo. 😉

A simple first pass solution may be just to raise if the chunks are irregular and request the user make them uniform.